bigquery_etl/routine/parse_routine.py

```diff
@@ -144,186 +158,162 @@ def from_text(text, dataset, name, filepath=None, description="", is_defined=Tru
     return {
-        raw_udf.name: raw_udf
-        for udf_dir in (udf_dirs or get_udf_dirs(UDF_DIRS) + list(MOZFUN_DIR))
-        for root, dirs, files in os.walk(udf_dir)
+        raw_routine.name: raw_routine
+        for root, dirs, files in os.walk(project_dir)
         if os.path.basename(root) != EXAMPLE_DIR
         for filename in files
         if filename in (UDF_FILE, PROCEDURE_FILE)
-        for raw_udf in (RawUdf.from_file(os.path.join(root, filename)),)
+        for raw_routine in (RawRoutine.from_file(os.path.join(root, filename)),)
     }
 
 
-def parse_udf_dirs(*udf_dirs):
-    """Read contents of udf_dirs into ParsedUdf instances."""
+def parse_routines(project_dir):
+    """Read routine contents of the project dir into ParsedRoutine instances."""
     # collect udfs to parse
-    raw_udfs = read_udf_dirs(*udf_dirs)
+    raw_routines = read_routine_dir(project_dir)
+    raw_routines.update(read_routine_dir(os.path.join(SQL_DIR, "mozfun")))
+    raw_routines.update(read_routine_dir(ASSERT_UDF_DIR))
 
     # prepend udf definitions to tests
-    for raw_udf in raw_udfs.values():
-        tests_full_sql = udf_tests_sql(raw_udf, raw_udfs)
-        yield ParsedUdf.from_raw(raw_udf, tests_full_sql)
+    for raw_routine in raw_routines.values():
+        tests_full_sql = routine_tests_sql(raw_routine, raw_routines, project_dir)
+        yield ParsedRoutine.from_raw(raw_routine, tests_full_sql)
 
 
-def accumulate_dependencies(deps, raw_udfs, udf_name):
+def accumulate_dependencies(deps, raw_routines, udf_name):
     """
-    Accumulate a list of dependent UDF names.
+    Accumulate a list of dependent routine names.
 
-    Given a dict of raw_udfs and a udf_name string, recurse into the
-    UDF's dependencies, adding the names to deps in depth-first order.
+    Given a dict of raw_routines and a udf_name string, recurse into the
+    routine's dependencies, adding the names to deps in depth-first order.
     """
-    if udf_name not in raw_udfs:
+    if udf_name not in raw_routines:
         return deps
 
-    raw_udf = raw_udfs[udf_name]
-    for dep in raw_udf.dependencies:
-        deps = accumulate_dependencies(deps, raw_udfs, dep)
+    raw_routine = raw_routines[udf_name]
+    for dep in raw_routine.dependencies:
+        deps = accumulate_dependencies(deps, raw_routines, dep)
     if udf_name in deps:
         return deps
     else:
         return deps + [udf_name]
 
 
-def udf_usages_in_file(filepath):
-    """Return a list of UDF names used in the provided SQL file."""
-    with open(filepath) as f:
-        text = f.read()
-    return udf_usages_in_text(text)
+def routine_usages_in_text(text, project):
+    """Return a list of routine names used in the provided SQL text."""
+    sql = sqlparse.format(text, strip_comments=True)
+    routines = get_routines(project)
 
+    udf_usages = []
+
+    for routine in routines:
+        if routine["name"] in sql:
+            udf_usages.append(routine["name"])
```

Comment:
This very much matches what we're doing up above, in the `RawRoutine.from_text` function. I'm wondering if we can instead use `RawRoutine.from_text(defined=False)` for these tests, like we do in the [hmac tests](https://github.com/mozilla/bigquery-etl/blob/master/tests/validation/hmac_sha256.py#L40).

Reply:
(Note: I don't think we should block this PR from landing on this change, since the problem existed before it).